Quarkus 1.3.0.Final gradle does not start with Undertow Servlet

**Describe the bug**
When I include Undertow Servlet in a Gradle project, quarkus application does not start.

**Expected behavior**
Quarkus app should start without throwing an error

**Actual behavior**
(Describe the actual behavior clearly and concisely.)


**To Reproduce**
Steps to reproduce the behavior:
1.  Go to https://code.quarkus.io/ 
2.  Change Build Tool to: Gradle
3.  Add Undertow Servlet as a dependency.
4. Generate project
5. Try running with quarkusDev build task.

Project throws java.lang.ExceptionInInitializerError

The same configuration when used with maven starts fine.


**Environment (please complete the following information):**
 - Output of `uname -a` or `ver`: 
 - Output of `java -version`: 
 - GraalVM version (if different from Java): 
 - Quarkus version or git rev: 
 - Build tool (ie. output of `mvnw --version` or `gradlew --version`): 

**Additional context**
(Add any other context about the problem here.)

I see this stacktrace, is the same you see? 

```
2020-03-19 17:04:51,293 INFO  [org.jbo.threads] (main) JBoss Threads version 3.0.1.Final
2020-03-19 17:04:51,934 ERROR [io.qua.dev.DevModeMain] (main) Failed to start Quarkus: java.lang.ExceptionInInitializerError
        at java.base/java.lang.Class.forName0(Native Method)
        at java.base/java.lang.Class.forName(Class.java:398)
        at io.quarkus.runner.bootstrap.StartupActionImpl.run(StartupActionImpl.java:80)
        at io.quarkus.dev.IsolatedDevModeMain.firstStart(IsolatedDevModeMain.java:60)
        at io.quarkus.dev.IsolatedDevModeMain.accept(IsolatedDevModeMain.java:236)
        at io.quarkus.dev.IsolatedDevModeMain.accept(IsolatedDevModeMain.java:39)
        at io.quarkus.bootstrap.app.CuratedApplication.runInCl(CuratedApplication.java:131)
        at io.quarkus.bootstrap.app.CuratedApplication.runInAugmentClassLoader(CuratedApplication.java:84)
        at io.quarkus.dev.DevModeMain.start(DevModeMain.java:108)
        at io.quarkus.dev.DevModeMain.main(DevModeMain.java:47)
Caused by: java.lang.RuntimeException: Failed to start quarkus
        at io.quarkus.runner.ApplicationImpl.<clinit>(ApplicationImpl.zig:400)
        ... 10 more
Caused by: java.util.ServiceConfigurationError: io.undertow.servlet.ServletExtension: io.undertow.websockets.jsr.Bootstrap not a subtype
        at java.base/java.util.ServiceLoader.fail(ServiceLoader.java:588)
        at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.hasNextService(ServiceLoader.java:1236)
        at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.hasNext(ServiceLoader.java:1264)
        at java.base/java.util.ServiceLoader$2.hasNext(ServiceLoader.java:1299)
        at java.base/java.util.ServiceLoader$3.hasNext(ServiceLoader.java:1384)
        at io.undertow.servlet.core.DeploymentManagerImpl.handleExtensions(DeploymentManagerImpl.java:276)
        at io.undertow.servlet.core.DeploymentManagerImpl.deploy(DeploymentManagerImpl.java:155)
        at io.quarkus.undertow.runtime.UndertowDeploymentRecorder.bootServletContainer(UndertowDeploymentRecorder.java:442)
        at io.quarkus.deployment.steps.UndertowBuildStep$build29.deploy_0(UndertowBuildStep$build29.zig:295)
        at io.quarkus.deployment.steps.UndertowBuildStep$build29.deploy(UndertowBuildStep$build29.zig:320)
        at io.quarkus.runner.ApplicationImpl.<clinit>(ApplicationImpl.zig:386)
        ... 10 more
```

--- a/independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/QuarkusClassLoader.java
+++ b/independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/QuarkusClassLoader.java
@@ -34,6 +34,7 @@ import org.objectweb.asm.ClassWriter;
 public class QuarkusClassLoader extends ClassLoader implements Closeable {
 
     private static final Logger log = Logger.getLogger(QuarkusClassLoader.class);
+    protected static final String META_INF_SERVICES = "META-INF/services/";
 
     static {
         registerAsParallelCapable();
@@ -132,6 +133,27 @@ public class QuarkusClassLoader extends ClassLoader implements Closeable {
         //        resources.add(element.getResource(nm).getUrl());
         //    }
         //}
+
+        //this is a big of a hack, but is nessesary to prevent service leakage
+        //in some situations (looking at you gradle) the parent can contain the same
+        //classes as the application. The parent aggregation stops this being a problem
+        //in most cases, however if there are no actual implementations of the service
+        //in the application then this can still cause problems
+        //this approach makes sure that we don't load services that would result
+        //in a ServiceConfigurationError
+        //see https://github.com/quarkusio/quarkus/issues/7996
+        if (name.startsWith(META_INF_SERVICES)) {
+            try {
+                Class<?> c = loadClass(name.substring(META_INF_SERVICES.length()));
+                if (c.getClassLoader() == this) {
+                    //if the service class is defined by this class loader then any resources that could be loaded
+                    //by the parent would have a different copy of the service class
+                    banned = true;
+                }
+            } catch (ClassNotFoundException ignored) {
+                //ignore
+            }
+        }
         for (ClassPathElement i : elements) {
             ClassPathResource res = i.getResource(nm);
             if (res != null) {

